File: lib/java/backport-util-concurrent-3.1-src/src/edu/emory/mathcs/backport/java/util/concurrent/atomic/AtomicLong.java
/*
 * Written by Doug Lea with assistance from members of JCP JSR-166
 * Expert Group and released to the public domain, as explained at
 * http://creativecommons.org/licenses/publicdomain
 */

package edu.emory.mathcs.backport.java.util.concurrent.atomic;

/**
 * A {@code long} value that may be updated atomically.  See the
 * {@link edu.emory.mathcs.backport.java.util.concurrent.atomic} package specification for
 * description of the properties of atomic variables. An
 * {@code AtomicLong} is used in applications such as atomically
 * incremented sequence numbers, and cannot be used as a replacement
 * for a {@link java.lang.Long}. However, this class does extend
 * {@code Number} to allow uniform access by tools and utilities that
 * deal with numerically-based classes.
 *
 * @since 1.5
 * @author Doug Lea
 */
public class AtomicLong extends Number implements java.io.Serializable {
    private static final long serialVersionUID = 1927816293512124184L;

    private volatile long value;

    /**
     * Creates a new AtomicLong with the given initial value.
     *
     * @param initialValue the initial value
     */
    public AtomicLong(long initialValue) {
        value = initialValue;
    }

    /**
     * Creates a new AtomicLong with initial value {@code 0}.
     */
    public AtomicLong() {
    }

    /**
     * Gets the current value.
     *
     * @return the current value
     */
    public final long get() {
        return value;
    }

    /**
     * Sets to the given value.
     *
     * @param newValue the new value
     */
    public final synchronized void set(long newValue) {
        value = newValue;
    }

    /**
     * Eventually sets to the given value.
     *
     * @param newValue the new value
     * @since 1.6
     */
    public final synchronized void lazySet(long newValue) {
        value = newValue;
    }

    /**
     * Atomically sets to the given value and returns the old value.
     *
     * @param newValue the new value
     * @return the previous value
     */
    public final synchronized long getAndSet(long newValue) {
        long old = value;
        value = newValue;
        return old;
    }

    /**
     * Atomically sets the value to the given updated value
     * if the current value {@code ==} the expected value.
     *
     * @param expect the expected value
     * @param update the new value
     * @return true if successful. False return indicates that
     * the actual value was not equal to the expected value.
     */
    public final synchronized boolean compareAndSet(long expect, long update) {
        if (value == expect) {
            value = update;
            return true;
        }
        else {
            return false;
        }
    }

    /**
     * Atomically sets the value to the given updated value
     * if the current value {@code ==} the expected value.
     *
     * <p>May <a href="package-summary.html#Spurious">fail spuriously</a>
     * and does not provide ordering guarantees, so is only rarely an
     * appropriate alternative to {@code compareAndSet}.
     *
     * @param expect the expected value
     * @param update the new value
     * @return true if successful.
     */
    public final synchronized boolean weakCompareAndSet(long expect, long update) {
        if (value == expect) {
            value = update;
            return true;
        }
        else {
            return false;
        }
    }

    /**
     * Atomically increments by one the current value.
     *
     * @return the previous value
     */
    public final synchronized long getAndIncrement() {
        return value++;
    }


    /**
     * Atomically decrements by one the current value.
     *
     * @return the previous value
     */
    public final synchronized long getAndDecrement() {
        return value--;
    }


    /**
     * Atomically adds the given value to the current value.
     *
     * @param delta the value to add
     * @return the previous value
     */
    public final synchronized long getAndAdd(long delta) {
        long old = value;
        value += delta;
        return old;
    }

    /**
     * Atomically increments by one the current value.
     *
     * @return the updated value
     */
    public final synchronized long incrementAndGet() {
        return ++value;
    }

    /**
     * Atomically decrements by one the current value.
     *
     * @return the updated value
     */
    public final synchronized long decrementAndGet() {
        return --value;
    }


    /**
     * Atomically adds the given value to the current value.
     *
     * @param delta the value to add
     * @return the updated value
     */
    public final synchronized long addAndGet(long delta) {
        return value += delta;
    }

    /**
     * Returns the String representation of the current value.
     * @return the String representation of the current value.
     */
    public String toString() {
        return Long.toString(get());
    }


    public int intValue() {
        return (int)get();
    }

    public long longValue() {
        return (long)get();
    }

    public float floatValue() {
        return (float)get();
    }

    public double doubleValue() {
        return (double)get();
    }

}

File: src/Spring/Spring.Threading/Threading/AtomicTypes/AtomicLong.cs
#region License

/*
 * Copyright 2002-2008 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#endregion

using System;
using System.Threading;

namespace Spring.Threading.AtomicTypes
{
	/// <summary> A <see cref="long"/> value that may be updated atomically.  See the
	/// An <see cref="AtomicLong"/> is used in applications such as atomically
	/// incremented sequence numbers, and cannot be used as a replacement
	/// for a <see cref="long"/>. 
	/// <p/>
	/// Based on the on the back port of JCP JSR-166.
	/// </summary>
	/// <author>Doug Lea</author>
	/// <author>Griffin Caprio (.NET)</author>
    /// <author>Andreas Doehring (.NET)</author>
	[Serializable]
    public class AtomicLong : IAtomic<long> //JDK_1_6
	{
		private long _longValue;

		/// <summary> 
		/// Creates a new <see cref="AtomicLong"/> with the given initial value.
		/// </summary>
		/// <param name="initialValue">
		/// The initial value
		/// </param>
		public AtomicLong(long initialValue)
		{
			_longValue = initialValue;
		}

		/// <summary> 
		/// Creates a new <see cref="AtomicLong"/> with initial value 0.
		/// </summary>
		public AtomicLong()
		{
		}

	    /// <summary> 
	    /// Atomically increments by one the current value.
	    /// </summary>
	    /// <returns> the previous value
	    /// </returns>
	    public long ReturnValueAndIncrement()
	    {
	        return Interlocked.Increment(ref _longValue) - 1;
	    }

	    /// <summary> 
	    /// Atomically decrements by one the current value.
	    /// </summary>
	    /// <returns> 
	    /// The previous value
	    /// </returns>
	    public long ReturnValueAndDecrement()
	    {
	        return Interlocked.Decrement(ref _longValue) + 1;
	    }

	    /// <summary> 
		/// Gets / Sets the current value.
		/// </summary>
		/// <returns> 
		/// The current value
		/// </returns>
		public long Value
		{
            get { return Interlocked.Read(ref _longValue); }
			set { Interlocked.Exchange(ref _longValue, value); }
		}

        /// <summary> 
		/// Eventually sets to the given value.
		/// </summary>
		/// <param name="newValue">
		/// The new value
		/// </param>
		public void LazySet(long newValue)
		{
            Interlocked.Exchange(ref _longValue, newValue);
		}
		/// <summary> 
		/// Atomically sets value to <paramref name="newValue"/> and returns the old value.
		/// </summary>
		/// <param name="newValue">
		/// The new value
		/// </param>
		/// <returns> 
		/// The previous value
		/// </returns>
		public long Exchange(long newValue)
		{
            return Interlocked.Exchange(ref _longValue, newValue);
		}
		/// <summary> 
		/// Atomically sets the value to <paramref name="newValue"/>
		/// if the current value == <paramref name="expectedValue"/>.
		/// </summary>
		/// <param name="expectedValue">
		/// The expected value
		/// </param>
		/// <param name="newValue">
		/// The new value
		/// </param>
		/// <returns> true if successful. False return indicates that
		/// the actual value was not equal to the expected value.
		/// </returns>
		public bool CompareAndSet(long expectedValue, long newValue)
		{
		    return Interlocked.CompareExchange(ref _longValue, newValue, expectedValue) == expectedValue;
		}

		/// <summary> 
		/// Atomically sets the value to <paramref name="newValue"/>
		/// if the current value == <paramref name="expectedValue"/>.
		/// May fail spuriously.
		/// </summary>
		/// <param name="expectedValue">
		/// The expected value
		/// </param>
		/// <param name="newValue">
		/// The new value
		/// </param>
		/// <returns> 
		/// True if successful.
		/// </returns>
		public virtual bool WeakCompareAndSet(long expectedValue, long newValue)
		{
            return Interlocked.CompareExchange(ref _longValue, newValue, expectedValue) == expectedValue;
        }

		/// <summary> 
		/// Atomically adds <paramref name="deltaValue"/> to the current value.
		/// </summary>
		/// <param name="deltaValue">
		/// The value to add
		/// </param>
		/// <returns> 
		/// The previous value
		/// </returns>
		public long AddDeltaAndReturnPreviousValue(long deltaValue)
		{
		    return Interlocked.Add(ref _longValue, deltaValue) - deltaValue;
		}
		/// <summary> 
		/// Atomically increments by one the current value.
		/// </summary>
		/// <returns> 
		/// The updated value
		/// </returns>
		public long IncrementValueAndReturn()
		{
		    return Interlocked.Increment(ref _longValue);
		}

		/// <summary> 
		/// Atomically decrements by one the current value.
		/// </summary>
		/// <returns> 
		/// The updated value
		/// </returns>
		public long DecrementValueAndReturn()
		{
		    return Interlocked.Decrement(ref _longValue);
		}

		/// <summary> 
		/// Atomically adds <paramref name="deltaValue"/> to the current value.
		/// </summary>
		/// <param name="deltaValue">
		/// The value to add
		/// </param>
		/// <returns> 
		/// The updated value
		/// </returns>
		public long AddDeltaAndReturnNewValue(long deltaValue)
		{
		    return Interlocked.Add(ref _longValue, deltaValue);
		}
		/// <summary> 
		/// Returns the String representation of the current value.
		/// </summary>
		/// <returns> 
		/// The String representation of the current value.
		/// </returns>
        public override string ToString() 
        {
			return Value.ToString();
		}

        /// <summary>
        /// Implicit converts <see cref="AtomicInteger"/> to int.
        /// </summary>
        /// <param name="atomicLong">
        /// Instance of <see cref="AtomicInteger"/>.
        /// </param>
        /// <returns>
        /// The converted int value of <paramref name="atomicLong"/>.
        /// </returns>
        public static implicit operator long(AtomicLong atomicLong)
        {
            return atomicLong.Value;
        }

	}
}